via/blocker.py

```diff
@@ -46,35 +48,159 @@ def __init__(self, application, checkmate_host=None):
+
+        if url.type == "via_landing_page":
+            return None, None, "landing_page"
+
+        if referrer.type in ("via_page", "via_sub_resource"):
+            return None, url, "sub_resource_check"
+
+        return url, None, "page_to_check"
+
+    def _check_url(self, url, allow_all=False):
+        """Check a URL with checkmate."""
 
         try:
-            hits = self._checkmate.check_url(url_to_annotate)
+            return self._checkmate.check_url(url, allow_all=allow_all)
+
         except CheckmateException as err:
             LOG.warning(
                 "Failed to check url against checkmate with error: {}".format(err)
             )
-            hits = None
+            return None
 
-        if hits:
-            template_name, status = self.templates.get(
-                hits.reason_codes[0], self.templates["other"]
-            )
+    def _render_block_template(self, hits, classified_url):
+        template_name, status = self.templates.get(
+            hits.reason_codes[0], self.templates["other"]
+        )
+
+        template = self._jinja_env.get_template(template_name).render(
+            url_to_annotate=classified_url.proxied_url,
+            domain_to_annotate=classified_url.proxied_domain,
+        )
+        return Response(template, status=status, mimetype="text/html")
 
-            template = self._jinja_env.get_template(template_name).render(
-                url_to_annotate=url_to_annotate,
-                domain_to_annotate=parsed_url.netloc,
+
+class ClassifiedURL(object):
+    """A URL with extra information about it's source."""
+
+    SUB_RESOURCE_RE = re.compile(r"^/([a-z]{2})_/(.*)$")
+
+    def __init__(self, type_, raw_url, proxied_url=None, resource_type=None):
+        """Create a URL with metadata.
+
+        :param type_: The classification of the URL
+        :param raw_url: The original raw URL
+        :param proxied_url: If this is a Via proxy, the proxied site
+        :param resource_type: If this is a "sub_resource" the resource type
+        """
+        self.raw_url = raw_url
+        self.type = type_
+
+        if proxied_url:
+            self.proxied_url, self.proxied_domain = self._clean_url(proxied_url)
+        else:
+            self.proxied_url, self.proxied_domain = None, None
+
+        if self.type == "via_sub_resource":
+            self.resource_type = resource_type
+        else:
+            self.resource_type = None
+
+    @classmethod
+    def _clean_url(cls, url):
+        """Clean up a URL to ensure it's fully qualified."""
+        url = url.lstrip("/")
+
+        parsed = urlparse(url)
+
+        if not parsed.scheme:
+            url = "http://" + url
+            parsed = urlparse(url)
+
+        return url, parsed.netloc
+
+    @classmethod
+    def classify(cls, raw_url, via_host, assume_via=False):
+        if not raw_url:
+            # Safety valve for being passed nonsense
+            return ClassifiedURL(None, raw_url)
```

Comment:
```suggestion
            return cls(None, raw_url)
```

It should either use `cls` or it should be a `staticmethod` not a `classmethod`. I've seen people say that, for factory methods, an `@classmethod` that uses `cls` is better because it avoids duplicating the class name. On the other hand an `@staticmethod` that uses `ClassifiedURL()` makes it easier to find all the places where `ClassifiedURL()` is called by grepping for it -- duplicating the class name may actually be a good thing, and deduplicating it may simply make the code harder to grep. (The reference to `cls.SUB_RESOURCE_RE` would also need to be changed.)

Reply:
I've kept this as a class method and used `cls`